src/wallet/wallet.cpp

```diff
@@ -422,6 +465,8 @@ bool CWallet::Unlock(const SecureString& strWalletPassphrase, bool accept_no_key
             if (!crypter.Decrypt(pMasterKey.second.vchCryptedKey, _vMasterKey))
                 continue; // try another master key
             if (CCryptoKeyStore::Unlock(_vMasterKey, accept_no_keys))
+                // Now that we've unlocked, upgrade the key metadata
+                UpgradeKeyMetadata();
```

Comment:
This means wallets will iterate through all key metadata every unlock?

Reply:
Perhaps a more efficient approach is to set a wallet flag once all metadata is upgraded?

Reply:
This is a point in favor of atomicity, then marking as complete.

Reply:
I've added a non-mandatory wallet flag which is set upon the upgrade's completion. This will make sure that the upgrade happens only once.